akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -210,6 +210,169 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
+              activeSubstreamsMap(nextElement._1).substreamSource.push(nextElement._2)
+              nextElement = null
+            } else tryPull(in)
+        }
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = fail(ex)
+
+      override def onDownstreamFinish(): Unit =
+        if (activeSubstreamsMap.isEmpty) completeStage() else setKeepGoing(true)
+
+      override def onPush(): Unit = {
+        val elem = grab(in)
+        val keyOption = getKeyOption(elem)
+
+        keyOption match {
+          case Some(key) ⇒ activeSubstreamsMap.get(key) match {
+            case Some(handler) ⇒
+              if (handler.substreamSource.isAvailable) handler.substreamSource.push(elem)
+              else nextElement = key -> elem
+            case None ⇒
+              if (activeSubstreamsMap.size == maxSubstreams)
+                fail(new IllegalStateException(s"Cannot open substream for key '$key': too many substreams open"))
+              else if (closedSubstreams(key) && !hasBeenPulled(in)) pull(in)
+              else runSubstream(key, elem)
+          }
+          case None ⇒ // do nothing
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (!completeWholeStage()) setKeepGoing(true)
+      }
+    })
+
+    private def runSubstream(key: K, value: T): Unit = {
+      val substreamSource = new SubSourceOutlet[T]("GroupBySource " + nextId)
+      val handler = new SubstreamHandler(key, value, substreamSource)
+
+      substreamSource.setHandler(handler)
+      activeSubstreamsMap += key -> handler
+
+      if (isAvailable(out)) {
+        push(out, Source.fromGraph(substreamSource.source))
+        scheduleOnce(key, timeout)
+        substreamPushed = None
+      } else {
+        setKeepGoing(true)
+        substreamPushed = Some(handler)
+      }
+    }
+
+    private def getKeyOption(elem: T): Option[K] = {
+      try {
+        Some(f(elem))
+      } catch {
+        case NonFatal(ex) ⇒
+          decider(ex) match {
+            case Supervision.Stop                         ⇒ fail(ex)
+            case Supervision.Resume | Supervision.Restart ⇒ if (!hasBeenPulled(in)) pull(in)
+          }
+          None
+      }
+    }
+
+    override protected def onTimer(timerKey: Any): Unit = {
+      activeSubstreamsMap.get(timerKey).foreach(handler ⇒ {
+        handler.substreamSource.timeout(timeout)
+        closedSubstreams += timerKey
+        activeSubstreamsMap -= timerKey
+      })
+      if (isClosed(in)) completeWholeStage()
+    }
+
+    private class SubstreamHandler(val key: K, val firstElement: T, val substreamSource: SubSourceOutlet[T]) extends OutHandler {
+      var firstPush: Boolean = true
+      def hasNextForSubSource = hasNextElement && nextElement._1 == key
+
+      private def completeSubStream(): Unit = {
+        substreamSource.complete()
+        activeSubstreamsMap -= key
+        closedSubstreams += key
+      }
+
+      private def checkForCompletion(): Unit = {
```

Comment:
Something like `completeHandler()` as it actually does complete rather than check for completion.